osu.Game/Rulesets/Objects/SliderPathExtensions.cs

```diff
@@ -25,6 +29,71 @@ public static void SnapTo<THitObject>(this THitObject hitObject, IDistanceSnapPr
+        {
+            var controlPoints = sliderPath.ControlPoints;
+            var originalControlPointTypes = controlPoints.Select(p => p.Type).ToArray();
+
+            controlPoints[0].Type ??= PathType.Linear;
+
+            // Inherited points after a linear point should be treated as linear points.
+            controlPoints.Where(p => sliderPath.PointsInSegment(p)[0].Type == PathType.Linear).ForEach(p => p.Type = PathType.Linear);
+
+            double[] segmentEnds = sliderPath.GetSegmentEnds().ToArray();
+            double[] distinctSegmentEnds = segmentEnds.Distinct().ToArray();
+
+            // Remove control points at the end which do not affect the visual slider path ("invisible" control points).
+            if (segmentEnds.Length >= 2 && Math.Abs(segmentEnds[^1] - segmentEnds[^2]) < 1e-10 && distinctSegmentEnds.Length > 1)
+            {
+                int numVisibleSegments = distinctSegmentEnds.Length - 2;
+                var nonInheritedControlPoints = controlPoints.Where(p => p.Type is not null).ToList();
+
+                int lastVisibleControlPointIndex = controlPoints.IndexOf(nonInheritedControlPoints[numVisibleSegments]);
+
+                // Make sure to include all inherited control points directly after the last visible non-inherited control point.
+                while (lastVisibleControlPointIndex + 1 < controlPoints.Count)
+                {
+                    lastVisibleControlPointIndex++;
+
+                    if (controlPoints[lastVisibleControlPointIndex].Type is not null)
+                        break;
+                }
+
+                // Remove all control points after the first invisible non-inherited control point.
+                controlPoints.RemoveRange(lastVisibleControlPointIndex + 1, controlPoints.Count - lastVisibleControlPointIndex - 1);
+            }
+
+            // Restore original control point types.
+            controlPoints.Zip(originalControlPointTypes).ForEach(x => x.First.Type = x.Second);
+
+            // Recalculate perfect curve at the end of the slider path.
+            if (controlPoints.Count >= 3 && controlPoints[^3].Type == PathType.PerfectCurve && controlPoints[^2].Type is null && distinctSegmentEnds.Length > 1)
+            {
+                double lastSegmentStart = distinctSegmentEnds[^2];
+                double lastSegmentEnd = distinctSegmentEnds[^1];
+
+                var oldCircleArcPath = new List<Vector2>();
+                sliderPath.GetPathToProgress(oldCircleArcPath, lastSegmentStart / lastSegmentEnd, 1);
+
+                var newCircleArcPoints = new[]
+                {
+                    oldCircleArcPath[0],
+                    oldCircleArcPath[oldCircleArcPath.Count / 2],
+                    oldCircleArcPath[^1]
+                };
+
+                var newCircleArcPath = PathApproximator.ApproximateCircularArc(newCircleArcPoints.AsSpan());
+                controlPoints[^2].Position = newCircleArcPath[newCircleArcPath.Count / 2];
```

Comment:
Why is the arc approximated again if you've already got the midpoint from the old circle arc?